File: coding-interview-bit-manipulation-plus/src/main/java/BitManipulation.java
```java
package com.example;

public class BitManipulation {

    /**
     * Counts the number of set bits in an integer.
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public int countSetBits(int n) {
        int count = 0;
        while (n > 0) {
            count += (n & 1);
            n >>= 1;
        }
        return count;
    }


    /**
     * Reverses the bits of an unsigned integer.
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public int reverseBits(int n) {
        int reversed = 0;
        for (int i = 0; i < 32; i++) {
            reversed <<= 1;
            reversed |= (n & 1);
            n >>= 1;
        }
        return reversed;
    }


    /**
     * Finds the single number that appears an odd number of times in an array.
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public int singleNumber(int[] nums) {
        int result = 0;
        for (int num : nums) {
            result ^= num;
        }
        return result;
    }


    /**
     * Swaps the odd and even bits of an integer.
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public int swapOddEvenBits(int n) {
        int evenBits = n & 0xAAAAAAAA; // Get even bits
        int oddBits = n & 0x55555555;  // Get odd bits
        return (evenBits >>> 1) | (oddBits << 1);
    }
}
```

File: coding-interview-bit-manipulation-pro/bitManipulation.js
/**
 * Reverses the bits of an unsigned integer.
 * @param {number} n The input integer.
 * @returns {number} The integer with reversed bits.
 */
const reverseBits = (n) => {
    let result = 0;
    for (let i = 0; i < 32; i++) {
        result <<= 1;
        result |= (n >> i) & 1;
    }
    return result >>> 0; // Ensure unsigned result
};

/**
 * Counts the number of set bits (1s) in an unsigned integer.
 * @param {number} n The input integer.
 * @returns {number} The number of set bits.
 */
const countSetBits = (n) => {
  let count = 0;
  while (n > 0) {
    count += (n & 1);
    n >>= 1;
  }
  return count;
};


/**
 * Finds the only non-repeating element in an array where every other element appears twice.
 * @param {number[]} nums The input array.
 * @returns {number} The single non-repeating element.
 */
const findSingleNonRepeating = (nums) => {
  let result = 0;
  for (let num of nums) {
    result ^= num;
  }
  return result;
};

/**
 * Checks if a number is a power of 2.
 * @param {number} n The input integer.
 * @returns {boolean} True if n is a power of 2, false otherwise.
 */
const isPowerOfTwo = (n) => {
  return (n > 0) && ((n & (n - 1)) === 0);
};

/**
 * Swaps two numbers without using a temporary variable.
 * @param {number} a First number.
 * @param {number} b Second number.
 * @returns {array} An array containing the swapped numbers.
 */
const swapNumbers = (a, b) => {
  a ^= b;
  b ^= a;
  a ^= b;
  return [a, b];
};


//Time and Space Complexity Analysis (Example for reverseBits)
//Time Complexity: O(log n) - We iterate through the bits of the number.
//Space Complexity: O(1) - Constant extra space is used.


module.exports = { reverseBits, countSetBits, findSingleNonRepeating, isPowerOfTwo, swapNumbers };